rust: Replace `keep_fingerprints` logic with unconditional `cargo clean`

ijc opened the pull request:
Due to Earthly's layer cache code added with `COPY` (even with `--keep-ts`) can
end up with timestamps (`mtime`) corresponding to the point of creation of the
cache entry, not the current time.
    
However on a following build the `target` mount cache may contain builds from
other branches, with different code for those dependencies, which have a newer
`mtime`. In this case `cargo` will think it can use the cached dependency
instead of rebuilding (because the code appears older than the cached entry
under `target`).
    
Avoid this by using `cargo clean` to remove the build artifacts for any local
crate. This should become unnecessary with https://github.com/rust-lang/cargo/issues/14136
    
This replaces the old behaviour of removing the fingerprints directory. Using
`cargo clean` uses a proper cargo API rather than relying on implementation
details like where the fingerprints live and what the consequence removing them
is. It may also keep the cached data smaller since it removes the build
artifacts which will likely never be reused due to the lack of fingerprint.
    
Note that the previous fingerprint cleaning was subject to a race where a
different parallel build could reintroduce some fingerprints between `DO
+REMOVE_SOURCE_FINGERPRINTS` and the `RUN ... cargo $args`. For that reason the
calls to `cargo clean` here are made within the same `RUN` command so that the
target cache remains locked.
    
By switching to `cargo metadata` the requirement for `tomljson` is removed.

---

This is an alternative to https://github.com/earthly/lib/pull/62.

I also included a couple of QOL changes which I think are useful.

ijc commented:
The downside of this change is that local crates must always be fully rebuilt. ie. caching is now only effective for 3rd party dependencies.

idelvall commented:
Interesting idea Ian. This is related to https://github.com/ijc/earthly-lib/tree/rust-avoid-caching-local-path-dependencies/rust#keep_fingerprints-false. 
Wondering if we should always remove the package fingerprints instead. 
What do you think?

ijc commented:
Hrm, yes, there is some overlap for sure.

My take would be that it is better to use formal APIs (`cargo clean`) rather than relying on implementation details like where the fingerprints live and what the consequence removing them is.

I would expect this also produces a space saving since I think removing the fingerprint doesn't remove the actual build artifacts, they will never be reused due to the lack of fingerprint so it is wasteful to leave them in the cache? Whereas clean will remove the actual binary blobs too.

ijc commented on rust/Earthfile:
```diff
@@ -61,11 +66,17 @@ CARGO:
+      echo "+CARGO: cargo $args"; \
       cargo $args; \
+      if [ "$EARTHLY_DO_NOT_CACHE_PATH_BASED_DEPENDENCIES" = "true" ]; then \
+          cargo metadata --format-version=1  --no-deps | jq -r '.packages[] | select(.id | startswith("path+file://")) | .name' | xargs -I{} cargo clean -p {}; \
```
This is too complicated and doesn't match "path" deps like it claims, it's just matching all local crates in an complicated way (since it is matching the crate location not something to do with the dep tree). It could be replaced with `.packages[].name` which is exactly the same due to `--no-deps`.

I think the path based dependencies thing is a red-herring, since we'd want to not cache any local crate since even the crate being built could suffer the false target cache hit problem, not only the things it depends on.

If the principal here is one we want to go fortward with then I'll update to simplify and rename/reword things.

idelvall commented:
> My take would be that it is better to use formal APIs (cargo clean) rather than relying on implementation details like where the fingerprints live and what the consequence removing them is.

Agree

>I would expect this also produces a space saving since I think removing the fingerprint doesn't remove the actual build artifacts, they will never be reused due to the lack of fingerprint so it is wasteful to leave them in the cache? Whereas clean will remove the actual binary blobs too.

`clean` sounds like the way to go, for sure

> I think the path based dependencies thing is a red-herring, since we'd want to not cache any local crate since even the crate being built could suffer the false target cache hit problem, not only the things it depends on.

What if we remove the fingerprints option and we always avoid caching path-based deps?

ijc commented:
> What if we remove the fingerprints option and we always avoid caching path-based deps?

I think that would be fine.

I'd probably leave an option to disable the clean still, but have it do the clean by default.

I'm not sure, but I think with this the `--keep-ts` stuff may no longer be needed, unless there is some other reason why it would be useful.

idelvall commented:
>I'm not sure, but I think with this the --keep-ts stuff may no longer be needed, unless there is some other reason why it would be useful.

I don't see a reason anymore neither

ijc commented:
> I'd probably leave an option to disable the clean still, but have it do the clean by default.

Actually, in trying to write a doc update as part of this I realised I couldn't really put into succinct words the circumstances where this option would be useful -- it's a pretty subtle case and my suspicion is that any attempt to use it would more than likely be wrong. Maybe I'll just make it unconditional.

ijc committed:
rust: Avoid `rename-output.sh` output if there is no output

This avoids producing:
```
+lint | no files found within ./target matching the provided output regexp
+lint | find: '/tmp/earthly/lib/rust': No such file or directory
```
when there is no output, because `copy-output.sh` only creates that path `if [
-n \"\$1\" ]`.

Apply the same condition to the `rename-output.sh` end of things.

ijc committed:
rust: Echo `cargo $args` into the logs

The logs otherwise just contain:
```
--> RUN set -e; cargo $args; cargo sweep -r -t $EARTHLY_SWEEP_DAYS; cargo sweep -r -i; $EARTHLY_FUNCTIONS_HOME/copy-output.sh "$output";
```

which doesn't show what is actually being run.

ijc committed:
rust: Replace `keep_fingerprints` logic with unconditional `cargo clean`

Due to Earthly's layer cache code added with `COPY` (even with `--keep-ts`) can
end up with timestamps (`mtime`) corresponding to the point of creation of the
cache entry, not the current time.

However on a following build the `target` mount cache may contain builds from
other branches, with different code for those dependencies, which have a newer
`mtime`. In this case `cargo` will think it can use the cached dependency
instead of rebuilding (because the code appears older than the cached entry
under `target`).

Avoid this by using `cargo clean` to remove the build artifacts for any local
crate. This should become unnecessary with https://github.com/rust-lang/cargo/issues/14136

This replaces the old behaviour of removing the fingerprints directory. Using
`cargo clean` uses a proper cargo API rather than relying on implementation
details like where the fingerprints live and what the consequence removing them
is. It may also keep the cached data smaller since it removes the build
artifacts which will likely never be reused due to the lack of fingerprint.

Note that the previous fingerprint cleaning was subject to a race where a
different parallel build could reintroduce some fingerprints between `DO
+REMOVE_SOURCE_FINGERPRINTS` and the `RUN ... cargo $args`. For that reason the
calls to `cargo clean` here are made within the same `RUN` command so that the
target cache remains locked.

By switching to `cargo metadata` the requirement for `tomljson` is removed.

ijc commented on rust/Earthfile:
```diff
@@ -56,16 +52,18 @@ CARGO:
+      echo "+CARGO: copying output"; \
+      $EARTHLY_FUNCTIONS_HOME/copy-output.sh "$output"; \
+      echo "+CARGO: removing local crates from target cache"; \
+      cargo metadata --format-version=1  --no-deps | /tmp/jq -r '.packages[].name' | xargs -I{} cargo clean -p {};
```
The call to `cargo clean` needs to match the profile used by `cargo $args` e.g. needs to include `--release` or `--profile=foo` as needed. Or I need to figure out a way to say "all profiles".

I'll have a think.

ijc replied:
I can't find a way to get a list of interesting profiles to clean. Other than "ask the user to supply it" or something ugly involving inspecting `$args` I can't currently see a way to make this approach work.

ijc replied:
This is further complicated by tools such as [`cargo llvm-cov`](https://github.com/taiki-e/cargo-llvm-cov) which overrides target dir (to `target/llvm-cov-target/`).
